### PR TITLE
Remove storage versions flag

### PR DESCRIPTION
**Pull request by caesarxuchao**
This pull follows up https://github.com/kubernetes/kubernetes/pull/68080. The `--storage-versions` flag of kube-apiserver is removed.

```release-note
The --storage-versions flag of kube-apiserver is removed. The storage versions will always be the default value built-in the kube-apiserver binary.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go [hunk 1]
```diff
@@ -39,10 +39,7 @@ type DefaultResourceEncodingConfig struct {
 }
 
 type GroupResourceEncodingConfig struct {
-	DefaultExternalEncoding   schema.GroupVersion
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go]: why remove the default group/version targets?

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go [hunk 2]
```diff
@@ -112,7 +100,7 @@ func (o *DefaultResourceEncodingConfig) InMemoryEncodingFor(resource schema.Grou
 
 	resourceOverride, resourceExists := groupEncoding.InternalResourceEncodings[resource.Resource]
 	if !resourceExists {
-		return groupEncoding.DefaultInternalEncoding, nil
+		return schema.GroupVersion{Group: resource.Group, Version: runtime.APIVersionInternal}, nil
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go]: why was this changed?

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go [hunk 3]
```diff
@@ -94,7 +81,8 @@ func (o *DefaultResourceEncodingConfig) StorageEncodingFor(resource schema.Group
 
 	resourceOverride, resourceExists := groupEncoding.ExternalResourceEncodings[resource.Resource]
 	if !resourceExists {
-		return groupEncoding.DefaultExternalEncoding, nil
+		// return the most preferred external version for the group
+		return o.scheme.PrioritizedVersionsForGroup(resource.Group)[0], nil
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go]: same here, didn't expect this

**Comment by liggitt** [comment]
the etcd_storage integration test would be the most interesting test to ensure still works after this

I also expected the flag to be announced/released as deprecated prior to removing it

**Comment by caesarxuchao** [comment]
I'll try to finish this PR and remove the "WIP" today, please do not review the details yet.

I'll take another look at the etcd_storage integration test.

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go (hunk 1 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go]: It's easier to understand if you take a look at the commit that removes these two fields: https://github.com/kubernetes/kubernetes/pull/67678/commits/a4d4e1e07c3a2711e8b9d8465dc1b83c28515e9e.

There were two ways to set the `DefaultExternalEncoding`.
1. via `SetVersionEncoding`, which was mainly called by tests. I removed all of them and tests still passed. It's also called by `MergeGroupEncodingConfigs`, which was dead code after I removed `--storage-configs`.
2. via `SetResourceEncoding`, which is used to set special encoding version for less stable resources in a group, like `cronjob` stays at `v1beta1` while the `batch` group prefers `v1`. Setting the group's default version to the special resource's version in this case is wrong, because the group's default version is and should be decided by `legacyscheme.Scheme.PreferredVersionAllGroups()`.

**Comment by liggitt** [comment]
> I'll try to finish this PR and remove the "WIP" today, please do not review the details yet.

if the first step is to mark the flag deprecated, then wait for the deprecation period, it seems like the first PR wouldn't be able to remove all of this yet.

**Comment by caesarxuchao** [comment]
> if the first step is to mark the flag deprecated

Can we keep the flag, but make it non-functional, and mark it as deprecated?

It shouldn't break regular users, unless someone reads etcd directly.

If we keep this flag functioning for two more releases, storage migration won't be safe unless users follow instructions to disable this flag.

**Comment by caesarxuchao** [comment]
Tests are passing, especially the `etcd_storage_path_test`. Removing WIP.

**Comment by caesarxuchao** [comment]
@deads2k @liggitt @lavalamp what do you think about https://github.com/kubernetes/kubernetes/pull/67678#issuecomment-415828665? If we can't make this flag non-functional, I can prepare another PR to announce its deprecation.

**Comment by enj** [comment]
@caesarxuchao I do not think it is valid to remove the flag's functionality without at least one release with an associated deprecation notice.

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go (hunk 1 above)
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go]: I vaguely remember @deads2k's saying something along the lines of: "each API type should define its storage version and there should no way to configure it otherwise."  But I may be misremembering.

The ETCD test appears to be passing so I guess this change is fine.

Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go]: My previous comment was rather about how the code was written than explaining why I made the change. I'll try to do the latter now.

Before this PR, there are three layers of "encoding version":
* the default encoding version is stored in the scheme,
* user can override the encoding version of a group via `--storage-version`, the value ends up in `DefaultResourceEncodingConfig.groups[<group>].DefaultExternalEncoding`,
* for less stable resource in a group, e.g., `cronjob` in the `batch` group, we hardcode the less stable resource to be encoded in a different version than the group, see [code](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubeapiserver/default_storage_factory_builder.go#L59-L61). This overriding value ends up in `DefaultResourceEncodingConfig.groups[<group>].ExternalResourceEncodings[<resource name>]`.

In this PR, we removed the `--storage-versions` flag, so the `DefaultExternalEncoding` is not needed. We only need to store the special encoding for those less stable types. This simplified the code in this file a lot.

File: hack/jenkins/test-dockerized.sh [hunk 4]
```diff
@@ -55,4 +55,3 @@ go install ./cmd/...
 
 make test-cmd
 make test-integration
-./hack/test-update-storage-objects.sh
```
Comment (caesarxuchao) [inline hack/jenkins/test-dockerized.sh]: We should remove the script as well. Tracked in https://github.com/kubernetes/kubernetes/issues/69299.

**Comment by caesarxuchao** [comment]
Rebased. @liggitt PTAL. Thanks.

**Comment by liggitt** [comment]
@caesarxuchao if you want to shorten the normal deprecation period (1 release or 6 months, whichever is longer), we need more compelling evidence that removal won't negatively affect people, or that keeping it functional for the documented deprecation period will cause harm

**Comment by caesarxuchao** [comment]
I can wait for 1.14.

**Comment by liggitt** [comment]
deprecation period is satisfied, can rebase and merge this for 1.14

**Comment by caesarxuchao** [comment]
I'm hesitant to remove the flag now because the storage version migrator needs the flag to run its e2e test, see this [line](https://github.com/kubernetes-sigs/kube-storage-version-migrator/blob/master/test/e2e/test-cmd.sh#L130). I can't figure out another way of setting up the test that's as flexible as using the --storage-version flag.

How about renaming the flag to `--storage-versions-test-only` instead of the removal?

**Comment by caesarxuchao** [comment]
kube-apiserver doesn't have testing only flags, but kube-controller-manager does have a few around the HostPath PV, e.g., enable-hostpath-provisioner.

I'll try one more workaround for the storage version migrator e2e test and report back.

**Comment by caesarxuchao** [comment]
Rebased. @liggitt PTAL, thanks.

**Comment by caesarxuchao** [comment]
It was successful enough, https://github.com/kubernetes-sigs/kube-storage-version-migrator/pull/16 :) 
So this pull was the same as it used to be, just a rebase.

**Comment by liggitt** [comment]
> It was successful enough, [kubernetes-sigs/kube-storage-version-migrator#16](https://github.com/kubernetes-sigs/kube-storage-version-migrator/pull/16) :)
> So this pull was the same as it used to be, just a rebase.

excellent, thanks

/lgtm
/approve

**Comment by liggitt** [comment]
/approve cancel
/lgtm cancel

actually, I'm not sure this got reviewed. will queue it up, glad it can progress for 1.14

**Comment by caesarxuchao** [comment]
No, it wasn't reviewed thoroughly.

I'll squash after it's reviewed.

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go [hunk 5]
```diff
@@ -34,86 +34,51 @@ type ResourceEncodingConfig interface {
...
 }
 
-type GroupResourceEncodingConfig struct {
-	DefaultExternalEncoding   schema.GroupVersion
-	ExternalResourceEncodings map[string]schema.GroupVersion
-
-	DefaultInternalEncoding   schema.GroupVersion
-	InternalResourceEncodings map[string]schema.GroupVersion
+type OverridingResourceEncoding struct {
+	ExternalResourceEncodings schema.GroupVersion
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go]: nit: ExternalResourceEncoding

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go [hunk 6]
```diff
@@ -34,86 +34,51 @@ type ResourceEncodingConfig interface {
...
+	InternalResourceEncodings schema.GroupVersion
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config.go]: nit: InternalResourceEncoding

**Review by liggitt — COMMENTED** [review]
one comment on the variable names. looks reasonable otherwise. leaning heavily on the etcd storage path test (no change needed there is a really good sign)

**Comment by caesarxuchao** [comment]
Fixed the nits. PTAL. Thanks.
